server/src/main/java/org/elasticsearch/search/profile/SearchProfileQueryPhaseResults.java

```diff
@@ -0,0 +1,123 @@
+        int size = in.readInt();
+        shardResults = new HashMap<>(size);
+
+        for (int i = 0; i < size; i++) {
+            String key = in.readString();
+            SearchProfileQueryPhaseResult shardResult = new SearchProfileQueryPhaseResult(in);
+            shardResults.put(key, shardResult);
+        }
+        shardResults = Collections.unmodifiableMap(shardResults);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeInt(shardResults.size());
+        for (Map.Entry<String, SearchProfileQueryPhaseResult> entry : shardResults.entrySet()) {
+            out.writeString(entry.getKey());
+            entry.getValue().writeTo(out);
+        }
+    }
+
+    /**
+     * Merge the profiling information from some fetch results into this
+     * profiling information.
+     */
+    public SearchProfileResults merge(Collection<? extends SearchPhaseResult> fetchResults) {
+        Map<String, SearchProfileShardResult> mergedShardResults = new HashMap<>(shardResults.size());
+        for (SearchPhaseResult r : fetchResults) {
+            FetchSearchResult fr = r.fetchResult();
+            String key = fr.getSearchShardTarget().toString();
+            SearchProfileQueryPhaseResult search = shardResults.get(key);
+            if (search == null) {
+                throw new IllegalStateException(
+                    "Profile returned fetch information for ["
+                        + key
+                        + "] but didn't return search information. Search keys were "
+                        + shardResults.keySet()
+                );
+            }
+            mergedShardResults.put(key, new SearchProfileShardResult(search, fr.profileResult()));
+        }
+        for (Map.Entry<String, SearchProfileQueryPhaseResult> e : shardResults.entrySet()) {
+            if (false == mergedShardResults.containsKey(e.getKey())) {
+                mergedShardResults.put(e.getKey(), new SearchProfileShardResult(e.getValue(), null));
+            }
+        }
+        return new SearchProfileResults(mergedShardResults);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        SearchProfileQueryPhaseResults other = (SearchProfileQueryPhaseResults) obj;
+        return shardResults.equals(other.shardResults);
+    }
+
+    @Override
+    public int hashCode() {
+        return shardResults.hashCode();
+    }
+
+    /**
+     * Helper method to convert Profiler into InternalProfileShardResults, which
+     * can be serialized to other nodes, emitted as JSON, etc.
+     *
+     * @param profilers
+     *            The {@link Profilers} to convert into results
+     * @return A {@link SearchProfileQueryPhaseResult} representing the results for this
+     *         shard
+     */
+    public static SearchProfileQueryPhaseResult buildShardResults(Profilers profilers) {
```

Comment:
Maybe this should live on `SearchProfileQueryPhaseResult` (singular).

Reply:
I was thinking of maybe moving it into it's call site. I don't think it really fit into its old spot either.

Reply:
I think I'll move it into `Profilers`.